Split migration

qnkhuat opened the pull request:
implement #34361 by splitting the current migration file into 2 files:
- 000_legacy_migrations.yaml containing migrations from the dawn of time to the last migration of 44
- 001_update_migrations.yaml containing all migrations from 45 till now
  - Added a new sqlFile migration with id `v00.00-000` to this file, this migration is used to bootstrap metabase to the state of v44. Think of this like instead of running `000_leagcy_migrations.yaml`, we use this sqlfile instead.

So how migration will play out in real life?
Scenario 1: Fresh instance
1. Migrate with the 001_update_migrations.yaml file
2. the first migration will be run is `v00.00-000`

Scenario 2: Upgrade from pre-45
1. Migrate using the 000_legacy_migrations.yaml file
2. Migrate with the 001_update_migrations.yaml file
3. migration `v00.00-000` will be skipped

Scenario 3: Upgrade from post-45
1. Migrate with the 001_update_migrations.yaml file
2. migration `v00.00-000` will be skipped

### Other changes
- I cleaned up the lint for migrations to remove the unstrict version, it was there primarily to check migrations that are old and didn't follow the "strict" rules.

## Technical notes

### How I generated these dump SQL for each db engines:
1. I faked the db by running only migrations in `000_legacy_migrations.yaml` which created the db as of 44
5. then for each dbms, I used the following commands to export its schema dump

```bash
# postgres
pg_dump --schema-only --no-owner --exclude-table databasechangelog --exclude-table databasechangeloglock metabase > resources/initialization/metabase_postgres.sql

# mysql
mysqldump -u root --no-data --skip-comments --ignore-table=metabase.databasechangelog --ignore-table=metabase.databasechangeloglock metabase > resources/initialization/metabase_mysql.sql

# h2 

select table_name from information_schema.tables where table_schema = 'PUBLIC' and table_name not like 'DATABASECHANGELOG%';

SCRIPT NODATA TO 'resources/initialization/metabase_h2.sql' TABLE COLLECTION, PERMISSIONS, COLLECTION_BOOKMARK, COMPUTATION_JOB, VIEW_LOG, DASHBOARD_FAVORITE, NATIVE_QUERY_SNIPPET, QRTZ_PAUSED_TRIGGER_GRPS, DATA_MIGRATIONS, DEPENDENCY, QRTZ_SIMPROP_TRIGGERS, DIMENSION, METRIC_IMPORTANT_FIELD, LABEL, SETTING, METABASE_TABLE, TIMELINE, QRTZ_FIRED_TRIGGERS, PERMISSIONS_GROUP_MEMBERSHIP, CORE_USER, COMPUTATION_JOB_RESULT, CARD_LABEL, QUERY, QUERY_EXECUTION, TIMELINE_EVENT, SECRET, CARD_BOOKMARK, PERMISSIONS_GROUP, PERSISTED_INFO, REVISION, GROUP_TABLE_ACCESS_POLICY, QRTZ_TRIGGERS, QUERY_CACHE, QRTZ_BLOB_TRIGGERS, DASHBOARDCARD_SERIES, DASHBOARD_BOOKMARK, QRTZ_CALENDARS, QRTZ_CRON_TRIGGERS, MODERATION_REVIEW, PULSE_CARD, QRTZ_SCHEDULER_STATE, BOOKMARK_ORDERING, REPORT_DASHBOARD, REPORT_DASHBOARDCARD, QRTZ_LOCKS, PULSE_CHANNEL_RECIPIENT, REPORT_CARDFAVORITE, CORE_SESSION, SEGMENT, METABASE_FIELDVALUES, PULSE_CHANNEL, REPORT_CARD, METABASE_FIELD, TASK_HISTORY, METRIC, APPLICATION_PERMISSIONS_REVISION, QRTZ_SIMPLE_TRIGGERS, PERMISSIONS_REVISION, LOGIN_HISTORY, COLLECTION_PERMISSION_GRAPH_REVISION, PULSE, QRTZ_JOB_DETAILS, METABASE_DATABASE, ACTIVITY;
```
6. You might notice the dump we have is not exactly the output of these commands
7. Some of the changes I made
	- Include some inserts as part of "initialization data for metabase". You can find these at the end of the dump file. All of these are permission-related
	- Remove constraints for all id columns to not start at 1, see this [commit](https://github.com/metabase/metabase/pull/34400/commits/c1edd4dc26ca98e0d49802a34f3acaa8745c3add) for example. we have this because the DB I used to dump actually contains some rows (permissions_group, permissions.. etc).
	- Also some driver-specific changes I'll comment in the file instead
	
### Why choose 45 as the cut-off?
2 reasons:
- we started offering rollback at 45
- in 46 we introduced custom migration using clojure, which can't be translated to sql, so we can't really go further than that

### Interesting number

This sped up startup time quite a lot, especially for MySQL which I got from 17.5s => 13s (~26% improvement!)

qnkhuat committed:
split the migrations file into 2

qnkhuat committed:
logic to decide which liquibase to choose on startup

qnkhuat committed:
can initialize for postgres and mysql

qnkhuat committed:
fix so test can run for pg and mysql

qnkhuat committed:
init for h2

qnkhuat committed:
tests should beable to run

qnkhuat committed:
comment about the necessary of extra migrations

qnkhuat committed:
fix some tests

qnkhuat committed:
makes sure h2 does not have RESTART WITH for auto generated IDs

qnkhuat committed:
Fix to makes test-migrations works

qnkhuat committed:
move data initialization into sql files itself

qnkhuat committed:
comment out the old schema migration tests

qnkhuat committed:
tests run successfully locally

qnkhuat committed:
check the name correctly

qnkhuat committed:
Fix the mysql utf8 test and remove starting id of some table for mysql

qnkhuat committed:
test fixes for mysql?

qnkhuat committed:
tidying

qnkhuat committed:
- Annotate the sql with initialization data
- makes the select set_config in postgres script works

qnkhuat committed:
use transaction to init db

qnkhuat committed:
remove the setting commands in postgres dump

qnkhuat commented on resources/initialization/metabase_postgres.sql:
the raw dump of postgres includes some setting like timeout stuffs, this is I think is good if you execute these dump using things like `pg_restore`, but we're using the same connection to run dump with the application, we shouldn't set it. So Removed them in https://github.com/metabase/metabase/pull/34400/commits/d13733700e9076ce32481ba5dba9cadcbf820830

qnkhuat committed:
does this fix for mysql?

qnkhuat committed:
should fix mysql for real

qnkhuat committed:
remove format-table name

qnkhuat committed:
fix e2e tests

qnkhuat commented on test/metabase/db/schema_migrations_test.clj:
removed all the migration tests for migration < 45 because they are no longer part of the main migration flow.

I could make it work and revert the deletion if someone think there is a good reason to keep those tests around.

qnkhuat commented on src/metabase/db/setup.clj:
```diff
@@ -52,7 +56,29 @@
     (throw (Exception. (trs "Database requires manual upgrade.")))))
 
-(s/defn migrate!
+(mu/defn ^:private initialize-db!
```
one problem with this is we might run into race condition if someone deploy metabase as multi-instances.

liquibase itself used lock to avoid that but we're running outside of liqubiase context here.

Should we worry about that?

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

piranha commented on test/metabase/db/schema_migrations_test/impl.clj:
```diff
@@ -196,11 +197,26 @@
+                 :up
+                 (do
+                  (log/debugf "Finding and running migrations between %s and %s (inclusive)" start-id (or end-id "end"))
+                  (run-migrations-in-range! conn [start-id end-id]))
```
`version` does not seem to be used here?

qnkhuat replied:
yes, it's only used for downgrade.

qnkhuat commented on src/metabase/db/setup.clj:
```diff
@@ -125,37 +152,47 @@
                  setting/*disable-cache*         true]
-         (verify-db-connection   db-type data-source)
+         (verify-db-connection db-type data-source)
+         (when (fresh-install? db-type data-source)
```
TODO: we might run into a race condition for this if someone start metabase as multi-instances

qnkhuat committed:
Add a test for 3 setup scenarios

qnkhuat committed:
Add tests and simplify the linter

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat commented on test/metabase/db/custom_migrations_test.clj:
removed all the migration tests for migration < 45 because they are no longer part of the main migration flow.

I could make it work and revert the deletion if someone thinks there is a good reason to keep those tests around.

qnkhuat commented on resources/initialization/metabase_postgres.sql:
the raw dump of postgres includes some setting like timeout stuffs, this is I think is good if you execute these dump using things like pg_restore, but we're using the same connection to run dump with the application, we shouldn't set it. So Removed them in https://github.com/metabase/metabase/commit/d13733700e9076ce32481ba5dba9cadcbf820830

qnkhuat committed:
makes validate-migrations public so lint could runs

qnkhuat commented on bin/lint-migrations-file/src/change/common.clj:
cleaned the linting for migration as now we don't need to lint for the old stuffs anymore.

qnkhuat committed:
test fixes

qnkhuat committed:
fix the table name for mysql when downgrade

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
fix lint migration tests

qnkhuat committed:
remove the unecessary downgrade migration to set databasechangelog because for older versions we'll run liquibase/consolidate-liquibase-changesets anw

qnkhuat committed:
Fix the linting failure

qnkhuat commented on resources/initialization/metabase_postgres.sql:
the raw dump of postgres includes some setting like timeout stuffs, this is I think is good if you execute these dump using things like pg_restore, but we're using the same connection to run dump with the application, we shouldn't set it. So Removed them in https://github.com/metabase/metabase/commit/d13733700e9076ce32481ba5dba9cadcbf820830

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
fix test failed to compile

qnkhuat committed:
fix failing tests for mysql because we used the wrong casing for databasechangelog table name

qnkhuat committed:
mysql should use port 3306 for testing

qnkhuat committed:
fix tests for mysql

qnkhuat committed:
test mysql on 3306

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
using liquibase sqlFile changset to initialize MB instead of executing the file manually

qnkhuat committed:
fix lint migration tests

qnkhuat commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -0,0 +1,3601 @@
+
+  - objectQuotingStrategy: QUOTE_ALL_OBJECTS
+
+  - changeSet:
```
this migration will only be executed on new instances

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
fix mysql test by using a more powerful regex

qnkhuat commented on test/metabase/db/schema_migrations_test.clj:
```diff
@@ -9,37 +9,27 @@
 
   See `metabase.db.schema-migrations-test.impl` for the implementation of this functionality."
```
removed all the migration tests for migration < 45 because they are no longer part of the main migration flow.

I could make it work and revert the deletion if someone think there is a good reason to keep those tests around.

qnkhuat commented on bin/lint-migrations-file/src/change/strict.clj:
```diff
@@ -1,12 +1,13 @@
 (ns change.strict
```
the unstrict versions are for old migrations, now we moved all of them to `000_legacy_migrations.yaml` we don't need to worry about it anymore

qnkhuat commented on test/metabase/db/schema_migrations_test/impl.clj:
```diff
@@ -196,11 +197,25 @@
-       (log/debugf "Finding and running migrations between %s and %s (inclusive)" start-id (or end-id "end"))
-       (run-migrations-in-range! conn [start-id end-id]))))
+    (run-migrations-in-range! conn ["v00.00-000" start-id] {:inclusive-end? false})
+    (letfn [(migrate
```
previously , in order to run `migrate! :down` in migration tests, people have to manually call `db.setup/migrate! :down 47`.

this change makes it so that you can call downgrade with the `migrate` binding when using with `test-migrations` macro.

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

tsmacdonald commented on bin/lint-migrations-file/src/lint_migrations_file.clj:
```diff
@@ -22,25 +20,17 @@
+    id))
 
 (defn- distinct-change-set-ids? [change-log]
   ;; there are actually two migration 32s, so that's the only exception we're allowing.
```
```suggestion
```

tsmacdonald commented on src/metabase/db/custom_migrations.clj:
```diff
@@ -162,10 +162,10 @@
-    (qs/delete-trigger scheduler (triggers/key "metabase.task.abandonment-emails.trigger"))
-    (qs/delete-job scheduler (jobs/key "metabase.task.abandonment-emails.job"))
-    (qs/shutdown scheduler)))
+   (qs/start scheduler)
+   (qs/delete-trigger scheduler (triggers/key "metabase.task.abandonment-emails.trigger"))
+   (qs/delete-job scheduler (jobs/key "metabase.task.abandonment-emails.job"))
+   (qs/shutdown scheduler)))
```
```suggestion
    (qs/start scheduler)
    (qs/delete-trigger scheduler (triggers/key "metabase.task.abandonment-emails.trigger"))
    (qs/delete-job scheduler (jobs/key "metabase.task.abandonment-emails.job"))
    (qs/shutdown scheduler)))
```

tsmacdonald commented on src/metabase/db/liquibase.clj:
```diff
@@ -49,7 +50,59 @@
+
+(defn table-exists?
+  "Check if a table exists."
+  [table-name data-source-or-conn]
```
I only see this called once (as a Connection). When would it be a data source?

tsmacdonald commented:
![image](https://github.com/metabase/metabase/assets/784417/560843f7-2a40-4d07-af0e-f965aeff702a)

tsmacdonald reviewed (approved):
It's a big change and I think more people should look at it, but :+1: from me once green. Looks like a minor issue with the CI failure

qnkhuat committed:
fix CI failed to run

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
fix unable to run cypress tests

qnkhuat committed:
still using mdb.setup/migrate...

qnkhuat committed:
fix missing imports

metamben commented on src/metabase/db/liquibase.clj:
```diff
@@ -242,30 +303,24 @@
 
-(s/defn consolidate-liquibase-changesets!
+(mu/defn consolidate-liquibase-changesets!
   "Consolidate all previous DB migrations so they come from single file.
```
It's not a single file anymore, is it?

qnkhuat replied:
updated https://github.com/metabase/metabase/pull/34400/commits/99ca116e8203d269e506c52db741913711d6516e

metamben commented on test/metabase/db/setup_test.clj:
```diff
@@ -36,3 +43,38 @@
+    (testing "can setup a fresh db"
+      (mt/with-temp-empty-app-db [_conn driver/*driver*]
+        (is (= :done
+               (mdb.setup/setup-db! driver/*driver* (mdb.connection/data-source) true)))))))
```
What about checking if the expected changes are (or the last expected change is) there?

metamben commented on test/metabase/db/setup_test.clj:
```diff
@@ -36,3 +43,38 @@
+             "v45"))
+
+      (is (= :done
+             (mdb.setup/setup-db! driver/*driver* (mdb.connection/data-source) true))))))
```
There is a fair amount of repetition here, might want to extract it.

qnkhuat replied:
how about this ? https://github.com/metabase/metabase/pull/34400/commits/dbedb44dfc63fd23c88d6b728c70d4d19d418c2a

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
Resolve Tamas's suggestions:
- liquibase_legacy.yaml specify exactly which file is included
- update various docs

qnkhuat committed:
kondo

qnkhuat committed:
use absolute id for setup tests

qnkhuat commented on resources/liquibase_legacy.yaml:
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - includeAll:
```
yes we should, I updated it in https://github.com/metabase/metabase/pull/34400/commits/99ca116e8203d269e506c52db741913711d6516e

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration

qnkhuat committed:
kondo

qnkhuat committed:
Merge branch 'master' into ngoc-split-migration
